AssertionError when setting up a large cluster

When setting up a connection to a large cluster, the client will throw a TimeoutException because it takes longer than expected for the server to allocate the cluster. Previously we could work around this by establishing the connection in a retry loop, but now the server will crash with an AssertionError if we try to reestablish the connection. (When setting up a small cluster we don't get a TimeoutException and so we don't get an AssertionError)

The exception thrown by the server:

com.tc.util.TCAssertionError: Assertion failed
at com.tc.util.Assert.failure(Assert.java:55)
at com.tc.util.Assert.eval(Assert.java:64)
at com.tc.util.Assert.assertTrue(Assert.java:85)
at com.tc.objectserver.handler.ProcessTransactionHandler$1.handleEvent(ProcessTransactionHandler.java:98)
at com.tc.objectserver.handler.ProcessTransactionHandler$1.handleEvent(ProcessTransactionHandler.java:91)
at com.tc.async.impl.StageQueueImpl$HandledContext.runWithHandler(StageQueueImpl.java:502)
at com.tc.async.impl.StageImpl$WorkerThread.run(StageImpl.java:197)

Client code to reproduce the exception (note the cluster size of "(64, MemoryUnit.GB)" - this size will throw the TimeoutException, but if we only allocate 8gb then we don't get the TimeoutException or the AssertionError):

    final URI clusterURI = URI.create("terracotta://localhost:9510/cachemanager");
    PersistentCacheManager cacheManager = null;

    final CacheManagerBuilder<PersistentCacheManager> clusteredCacheManagerBuilder =
        CacheManagerBuilder.newCacheManagerBuilder()
            .with(ClusteringServiceConfigurationBuilder.cluster(clusterURI)
                .autoCreate().defaultServerResource("primary-server-resource")
            )
            .withCache("clustered-cache",   CacheConfigurationBuilder.newCacheConfigurationBuilder(Long.class, byte[].class,
              ResourcePoolsBuilder.newResourcePoolsBuilder()
                  .with(ClusteredResourcePoolBuilder.clusteredDedicated(64, MemoryUnit.GB))));
    while (true) {
      try {
        cacheManager = clusteredCacheManagerBuilder.build(true); // this throws the TimeoutException
        break;
      } catch (Exception e) {
        e.printStackTrace();
      }
    }

--- a/dso-l2/src/main/java/com/tc/objectserver/handler/ProcessTransactionHandler.java
+++ b/dso-l2/src/main/java/com/tc/objectserver/handler/ProcessTransactionHandler.java
@@ -91,11 +91,15 @@ public class ProcessTransactionHandler {
   private final AbstractEventHandler<TCMessage> multiSender = new AbstractEventHandler<TCMessage>() {
     @Override
     public void handleEvent(TCMessage context) throws EventHandlerException {
-      invokeReturn.remove((ClientID)context.getDestinationNodeID(), context);
-      Assert.assertTrue(context.send());
-      if (LOGGER.isDebugEnabled()) {
+      NodeID destinationID = context.getDestinationNodeID();
+      invokeReturn.remove((ClientID)destinationID, context);
+      boolean didSend = context.send();
+      if (!didSend) {
+        // It is possible for this send to fail.  Typically, it means that the client has disconnected.
+        LOGGER.warn("Failed to send message to: " + destinationID);
+      } else if (LOGGER.isDebugEnabled()) {
         LOGGER.debug("sent " + context);
-      }      
+      }
     }
   };
   public AbstractEventHandler<TCMessage> getMultiResponseSender() {

